fah_xchem/analysis/diffnet.py

```diff
@@ -0,0 +1,311 @@
+import logging
+from math import log
+from typing import List, Optional
+
+import networkx as nx
+import numpy as np
+
+from ..schema import (
+    Compound,
+    CompoundAnalysis,
+    CompoundMicrostate,
+    Microstate,
+    MicrostateAnalysis,
+    PointEstimate,
+    TransformationAnalysis,
+)
+from .exceptions import AnalysisError, ConfigurationError, InsufficientDataError
+
+
+def pIC50_to_DG(pIC50: float, s_conc: float = 375e-9, Km: float = 40e-6) -> float:
+    """
+    Converts IC50 (in M units) to DG
+    Parameters
+    ----------
+    pIC50 : float
+        pIC50
+    s_conc : float, default=375E-9
+        Substrate concentration in M
+    Km : float, default=40E-6
+        Substrate concentration for half-maximal enzyme activity
+    Returns
+    -------
+    type
+        Description of returned object.
+    """
+    ic50 = 10 ** -pIC50
+
+    if ic50 > 1e-5:
+        logging.warning("Expecting IC50 in M units. Please check.")
+
+    Ki = ic50 / (1 + s_conc / Km)
+    return log(Ki)
+
+
+def get_compound_free_energy(microstates: List[MicrostateAnalysis]) -> PointEstimate:
+    r"""
+    Compute compound free energy from microstate free energies and
+    microstate free energy penalties.
+
+    The sum over microstates is given by
+
+    .. math:: g_c = - \log \sum_{i} \exp \left[-(s_{ci} + g_{ci}) \right]
+
+    Parameters
+    ----------
+    microstates : list of MicrostateAnalysis
+        Microstate free energies
+    """
+    penalized_free_energies = [
+        microstate.free_energy + microstate.microstate.free_energy_penalty
+        for microstate in microstates
+        if microstate.free_energy is not None
+    ]
+
+    if not penalized_free_energies:
+        raise InsufficientDataError("no microstate free energy estimates")
+
+    g = np.array([x.point for x in penalized_free_energies])
+    stderr = np.array([x.stderr for x in penalized_free_energies])
+
+    x = np.exp(-g)
+    z = np.sum(x)
+    gc = np.log(z)
+    dgc = -x / z
+
+    return PointEstimate(point=gc, stderr=np.sqrt(np.sum((dgc * stderr) ** 2)))
+
+
+def _validate_inputs(
+    compounds: List[Compound], transformations: List[TransformationAnalysis]
+):
+    # Microstates appearing as initial or final in transformations
+    transformation_microstates = set(
+        microstate
+        for transformation in transformations
+        for microstate in [
+            transformation.transformation.initial_microstate,
```

Comment:
`transformation.transformation`?

Reply:
Yes, same explanation as https://github.com/choderalab/fah-xchem/pull/77/files#r492231144 (and same potential fix)